README.md
[![crates.io](https://img.shields.io/crates/v/grafen.svg)](https://crates.io/crates/grafen) [![docs.rs](https://img.shields.io/badge/docs.rs-documentation-orange.svg)](https://docs.rs/crate/grafen) [![Build Status](https://travis-ci.org/pjohansson/grafen.svg?branch=master)](https://travis-ci.org/pjohansson/grafen)

Create graphene and other substrates or system configurations for use in molecular dynamics simulations.

This is a pet project to help me set up simulation systems for my research. It is focused on formats used by [Gromacs](http://www.gromacs.org/).

**As of 0.10 requires Rust Nightly for `nll` and `underscore_lifetimes` features.**

# Usage
```
USAGE:
    grafen [OPTIONS]

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

OPTIONS:
    -d, --database <database>      Path to residue and component database
    -c, --conf <input_confs>...
            Path to input configuration files to add as components

    -o, --output <output>
            Output configuration file [default: conf.gro]

    -t, --title <title>            Title of output system
```

# Available Substrates
Substrate definitions are read from a JSON database. An example is provided
in `assets/database.json`. This database contains a few residue
and the following substrate definitions:

## Graphene
A monolayer of carbon atoms set in a hexagonal honeycomb structure.
The spacing between every atom is 0.142 nm.

## Silica
A monolayer of rigid SiO2 molecules set in a triclinic formation with
with spacing 0.450 nm along both base vectors and an angle of 60 degrees
between them.

## Graphene Nanotube
A cylinder constructed of carbon atoms in the same structure as the above graphene.

# Configuration Files
The program supports reading configurations from disk and manipulating them in some ways. Currently read configurations can be extended by duplicating and cutting them, or cut into cylinders.

Such an example is included in the `database.json` file.

# Database
## Paths to Configuration Files
The database which is saved to disk can contain references to system configurations that it will read data from. If these paths are entered into the database as relative paths, they will be read as relative to the database location.

## Database Location
The program by default tries to read a database from disk. On *Linux* (and other non-OSX *unix* systems) it looks in a subdirectory to the locations specified by the `XDG_DATA_HOME` and `XDG_DATA_DIRS` (read-only) environment variables, or the `$USER/.local/share` directory. On *OSX* it looks in the same `XDG`-spec locations but also in the user and root `Library/Application Support` directories. On *Windows* in the directory set by the `APPDATA` environment variable.

# Development
See the [documentation](https://docs.rs/crate/grafen).

# License
The program is unlicensed. See [unlicense.org](http://unlicense.org) for details.


CHANGELOG.md
0.10
===
* Add the ability to read configuration files (currently only .gro) as components.
* Add a `Cuboid` component that is a surface, not a filled volume.
* A default database is used to store data in. It is located in a platform-dependent source directory and is read if no database is supplied as a CLI argument.

0.9
===
* Rewrite of the object construction framework. Instead of a messy matching to different object types using `IntoComponent`, the simple trait `Component` has been redone to yield all necessary information from objects.
* In particular `Component` prescribed the method `iter_atoms` which yields all atoms in a component with their indices as required for output to disk.
* Objects are added to the `DataBase` using a static dispatcher enum as a wrapper. This wrapper lets all objects exist in one `Vec`. It is created using the new macro `create_entry_wrapper` which passes through all the required implementations.
* Object definitions now contain all the information required to process and use them. Instead of containing as little information as possible and then have a separate object wrap them with more methods for `DataBase` use, etc. This second layer is unnecessarily confusing (although we still need one separate layer for the static dispatch, but as per above this is fully transparent).
* The `ResidueBase` object has been renamed to `Residue`.
* `System` has been moved to `system` and improved to user proper methods for yielding information for output. This includes implementing `iter_atoms` through the `Component` trait which links to all components` methods of that name and merges them to a single whole.

0.9.1
-----
* Add the ability to edit constructed components by translating them or cutting coordinates from them using another constructed component.
* Sheets are now constructed with periodic boundary conditions being applied. Residue coordinates will not extend beyond the component box size (although atoms belonging to the residue might).
* Sheets can now be constructed with their normal vector pointing along any axis.
* The user menus have been rewritten using a macro to ensure more consistent behaviour.

0.8
===
* The user interface has been reworked, using `dialoguer` instead of plain text selection.

0.7
===
* The command line interface has been reworked to allow for new types of objects.
* This includes the database being able to store these on disk.
* In particular, a `Cylinder` object has been added. It is a sheet of molecules, folded into a cylinder. Practically it represents (eg.) a nanotube.
* `Component`s now own their `ResidueBase`. As such the `Residue` object has been removed.

0.7.1
-----
* `Cylinder`s can now be created as filled volumes, not only sheets.

0.6
===
* Substrate definitions are now read from a database at runtime. This database is JSON formatted and implemented using the `serde` framework.
* The binary has been updated to construct systems as read from this database. This is an early implementation of the binary, it is in need of improvement.

0.6.1
-----
* It is now possible to properly edit the database by adding or removing residue and substrate definitions.
* A database can be saved and moved to different locations.
* The GROMOS file output has been changed to number atoms not per-residue but by their absolute index in the system.

0.5
===
* Substrate configurations are now user-defined instead of hard-coded in the library. They are made available through the `SubstrateConf` struct which is passed to the construction function.
* The base `System` struct has been moved along with its constituent `Residue` and `Atom` structs to a new module. This is in preparation for making the library more general and not only for constructing substrates.
* The `substrates` module tests have been cleaned up. A couple of unit tests remain for very specific functionality, but tests for creating a substrate has been moved to the `tests` folder as proper integration tests.

0.5.1
-----
* The `substrates` module has been renamed to the singular `substrate`.
* The `lattice` module is now a *private* submodule to `substrate`.
* The `output` module has been moved to the binary.
* A macro `resbase` has been added to easily construct `ResidueBase` objects.
* `ResidueBase` no longer implements methods for getting a silica or graphene base. These definitions belong in the binary, not in the library.

0.5.2
-----
* Implement a Poisson Disc distribution for generating substrates.

0.4
===
* Library and CLI utility split into separate packages
* Silica and graphene substrates implemented

0.4.1
-----
No notable changes.

0.4.2
-----
* Substrate residues can be shifted along z by a uniform random distribution
* This is added to the CLI utility as an option alongside an input position for the substrate


src/main.rs
//! Create graphene and other substrates for use in molecular dynamics simulations.
//!
//! # Usage
//! ```
//! USAGE:
//!     grafen [OPTIONS] <PATH> <X> <Y>
//!
//! FLAGS:
//!     -h, --help       Prints help information
//!     -V, --version    Prints version information
//!
//! OPTIONS:
//!     -t, --title <STR>    Title of system
//!
//! ARGS:
//!     <PATH>    Output GROMOS file (the extension will be set to .gro)
//!     <X>       Size along x
//!     <Y>       Size along y
//! ```
//!
//! # Available Substrates
//! Spacings and translations for all substrates is currently hard-coded.
//! Preferably this should be set in some configuration files or as an option
//! input by the user during runtime.
//!
//! ## Graphene
//! A monolayer of carbon atoms set in a hexagonal honeycomb structure.
//! The spacing between every atom is 0.142 nm.
//!
//! ## Silica
//! A monolayer of rigid SiO2 molecules set in a triclinic formation with
//! with spacing 0.450 nm along both base vectors and an angle of 60 degrees
//! between them.

extern crate ansi_term;
#[macro_use]
extern crate clap;

pub mod lattice;
pub mod config;
pub mod output;
pub mod substrates;

use config::Config;

use std::io;
use std::io::Write;
use std::process;

fn main() {
    let matches = clap_app!(grafen =>
        (version: crate_version!())
        (author: crate_authors!())
        (about: crate_description!())
        (@arg output: <PATH> "Output GROMOS file (the extension will be set to .gro)")
        (@arg x: <X> "Size along x")
        (@arg y: <Y> "Size along y")
        (@arg title: -t --title [STR] +takes_value "Title of system")
    ).get_matches();

    if let Err(err) = Config::new(matches).and_then(|conf| conf.run()) {
        let mut stderr = io::stderr();
        writeln!(&mut stderr, "{}", err).expect("could not write to stderr");
        process::exit(1);
    }
}
